src/autopas/tuning/tuningStrategy/ruleBasedTuning/RuleBasedTuning.h

```diff
@@ -0,0 +1,171 @@
+ public:
+  /**
+   * A function type used to print errors found in verify mode.
+   */
+  using PrintTuningErrorFunType =
+      std::function<void(const rule_syntax::ConfigurationOrder &order, const Configuration &actualBetterConfig,
+                         unsigned long betterRuntime, const Configuration &shouldBeBetterConfig,
+                         unsigned long shouldBeBetterRuntime, const LiveInfo &liveInfo)>;
+
+  /**
+   * Constructs a RuleBasedTuning strategy.
+   * @param searchSpace Set of all allowed configurations.
+   * @param verifyModeEnabled If verify mode should be enabled. False by default.
+   * @param ruleFileName The name of the file where the rules are stored.
+   * @param tuningErrorPrinter The function to call in verify mode if errors are found.
+   */
+  explicit RuleBasedTuning(const std::set<Configuration> &searchSpace, bool verifyModeEnabled = false,
+                           std::string ruleFileName = "tuningRules.rule",
+                           PrintTuningErrorFunType tuningErrorPrinter = {});
+
+  TuningStrategyOption getOptionType() override;
+
+  bool needsLiveInfo() const override;
+
+  void receiveLiveInfo(const LiveInfo &info) override;
+
+  void addEvidence(const Configuration &configuration, const Evidence &evidence) override;
+
+  void reset(size_t iteration, size_t tuningPhase, std::vector<Configuration> &configQueue,
+             const autopas::EvidenceCollection &evidenceCollection) override;
+
+  /**
+   * @returns in verify mode the summed up time that would have been skipped if verify mode was disabled and
+   * configurations would have been skipped due to the rules in the last tuning phase.
+   */
+  [[nodiscard]] long getLifetimeWouldHaveSkippedTuningTime() const;
+
+  /**
+   * @returns the summed up time of all configurations that have been tested by this tuning strategy.
+   */
+  [[nodiscard]] long getLifetimeTuningTime() const;
+
+  void optimizeSuggestions(std::vector<Configuration> &configQueue,
+                           const EvidenceCollection &evidenceCollection) override;
+
+ private:
+  /**
+   * Creates a multi-line string representation of all rules contained in the given rule file.
+   *
+   * Works by printing all non-empty, non-comment lines in the file.
+   *
+   * @param filePath Path to the file whose rules should be printed.
+   * @return String representation of all rules.
+   */
+  std::string rulesToString(const std::string &filePath) const;
+
+  /**
+   * Goes through all applicable configuration orders and checks if the result of the current configuration contradicts
+   * any rules when comparing with previously tested configurations in this tuning phase. If yes, calls
+   * tuningErrorPrinter.
+   *
+   * @param configuration
+   */
+  void verifyCurrentConfigTime(const Configuration &configuration) const;
+
+  /**
+   * Executes the rule file for the current simulation state. Puts all known live info as defines in front of the
+   * program.
```

Comment:
The second sentence is not particularly clear.

Reply:
Not really. What is meant by "definition of variables" i.e. what variables? Similarly, what is meant by program?

Reply:
fbf2c6563